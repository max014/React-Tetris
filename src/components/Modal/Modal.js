import React, { Component } from 'react';
import { connect } from 'react-redux';
import * as actionTypes from '../../store/actions/actionTypes';
import * as utility from '../../utility';
import styles from './Modal.module.css';
import * as actions from '../../store/actions/scoreBoard';

export class Modal extends Component {
	state = {
		highScore: false,
		name: "name"
	}

	componentDidMount() {
		if(utility.isHighScore(this.props.score, this.props.scores)){
			this.setState({highScore: true});
		}
	}

	handleChange = event => this.setState({name: event.target.value});

	postAndReset = () => {
		this.props.postScore(this.props.score, this.state.name);
		const extraScores = utility.pruneScores(this.props.scores);
		if(extraScores){
			extraScores.forEach(score => {
				this.props.deleteScore(score.id);
			});
		}
		this.props.startGame();
	}

	render() {
		let input = null;
		let button = <div className={styles.button} onClick={this.props.startGame}>ReStart</div>;

		if(this.state.highScore){
			input = (
				<div>
					<h4>You made the board!</h4>
					<span>Name:</span>
					<input
						maxlength="15"
						type="text"
						name="name"
						value={this.state.name}
						onChange={this.handleChange}/>
				</div>
			);
			button = <div className={styles.button} id="btn-post" onClick={this.postAndReset}>Post High Score</div>;
		}

		return (
			<div className={styles.Modal}>
				<h1>Score: {this.props.score}</h1>
				{input}
				{button}
			</div>
		);
	}
}

const mapStateToProps = state => {
    return {
    	scores: state.scores,
        score: state.score
    };
}

const mapDispatchToProps = dispatch => {
    return {
		postScore: (score, name) => dispatch(actions.postScore(score, name)),
		deleteScore: (id) => dispatch(actions.deleteScore(id)),
        startGame: () => dispatch({type: actionTypes.START_GAME})
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(Modal);