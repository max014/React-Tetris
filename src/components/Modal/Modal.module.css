.Modal{
	position: absolute;
	width: 35vw;
	left: 50%;
	top: 30%;
	background-color: black;
	border: 4px solid white;
	transform: translateX(-50%);
}
.Modal h1,
.Modal h4{
	color: white;
	text-align: center;
}
.Modal .button{
	text-align: center;
	background-color: #ccc;
	margin-left: 20px;
	margin-right: 20px;
	margin-bottom: 20px;
	font-size: 20px;
	cursor: pointer;
}

.Modal span{
	margin-left: 20px;
	color: white;
}

.Modal input{
	text-align: center;
	margin-left: 20px;
	margin-right: 20px;
	margin-bottom: 20px;
	font-size: 20px;
}