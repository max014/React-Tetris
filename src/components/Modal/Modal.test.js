import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Modal } from './Modal';

describe('Modal', () => {
    const postScore = jest.fn();
	const deleteScore = jest.fn();
    const startGame = jest.fn();
    const props = {
        scores: [
            {"_id":"5d38cc4c32558d73ff83608e","name":"a","score":1548980},
            {"_id":"5d38cd1032558d78cb836093","name":"Samsquanch","score":416800},
            {"_id":"5d38cd3532558d39b6836095","name":"Moby Dick","score":596480},
            {"_id":"5d38cd4732558d3a4c836096","name":"Crazy Larry","score":447720},
            {"_id":"5d38cd5f32558d6da5836097","name":"Morgan Freeman","score":1658640},
            {"_id":"5d38cd7032558de28f836098","name":"The Great Westifer","score":1021140},
            {"_id":"5d38cc4c32558d73ff83608e","name":"a","score":1548980},
            {"_id":"5d38cd1032558d78cb836093","name":"Samsquanch","score":416800},
            {"_id":"5d38cd3532558d39b6836095","name":"Moby Dick","score":596480},
            {"_id":"5d38cd4732558d3a4c836096","name":"Crazy Larry","score":447720},
            {"_id":"5d38cd5f32558d6da5836097","name":"Morgan Freeman","score":1658640}
        ],
        score: 500000,
        postScore,
        deleteScore,
        startGame
    }
    const modal = shallow(<Modal {...props}/>);
    it('renders properly', () => {
        expect(toJson(modal)).toMatchSnapshot();
    });

    it('calls `postScore` and `startGame` from props', () => {
        modal.state().highScore = true;
        modal.update();
        modal.find('#btn-post').simulate('click');
        expect(postScore).toHaveBeenCalledTimes(1);
        expect(deleteScore).toHaveBeenCalledTimes(1);
        expect(startGame).toHaveBeenCalledTimes(1);
    });

    it('sets `state.name`', () => {
        const name = 'max';
        modal.find('input').simulate('change', {target: {value: name}});
        expect(modal.state().name).toEqual(name);
    });
});