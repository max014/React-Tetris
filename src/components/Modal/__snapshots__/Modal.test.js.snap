// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renders properly 1`] = `
<div
  className="Modal"
>
  <h1>
    Score: 
    500000
  </h1>
  <div>
    <h4>
      You made the board!
    </h4>
    <span>
      Name:
    </span>
    <input
      maxlength="15"
      name="name"
      onChange={[Function]}
      type="text"
      value="name"
    />
  </div>
  <div
    className="button"
    id="btn-post"
    onClick={[Function]}
  >
    Post High Score
  </div>
</div>
`;
